File: 72.py
class Solution(object):
    def minDistance(self, word1, word2):
        """
        :type word1: str
        :type word2: str
        :rtype: int
        """
        result = []
        length1 = len(word1) + 1
        length2 = len(word2) + 1
        for i in range(length1):
            tmp_result = [0]  * length2
            result.append(tmp_result)
        for i in range(length1):
            result[i][0] = i
        for j in range(1, length2):
            result[0][j] = j
        for i in range(1, length1):
            for j in range(1, length2):
                if word1[i-1] == word2[j-1]:
                    result[i][j] = result[i-1][j-1]
                else:
                    result[i][j] = min(result[i-1][j-1], result[i-1][j], result[i][j-1]) + 1
        return result[-1][-1]

File: 72.cpp
class Solution {
public:
    int minDistance(string word1, string word2) {
        int m = word1.size();
        int n = word2.size();
        vector<vector<int>> p(m+1, vector<int>(n+1));
        for (int i = 0; i <= m;i++){
            p[i][0] = i;
        }
        for (int j = 0; j <= n;j++){
            p[0][j] = j;
        }
        for (int i = 1; i<= m; i++){
            for (int j = 1; j <= n; j++){
                if (word1[i-1] == word2[j-1]){
                    p[i][j] = p[i-1][j-1];
                }
                else{
                    p[i][j] = 1 + min({p[i-1][j], p[i][j-1], p[i-1][j-1]});
                }
            }
        }
        return p[m][n];
    }
};
